src/libcore/hint.rs

```diff
@@ -91,3 +91,42 @@ pub fn spin_loop() {
         }
     }
 }
+
+/// A function that is opaque to the optimizer, to allow benchmarks to
+/// pretend to use outputs to assist in avoiding dead-code
+/// elimination.
+///
+/// This function is a no-op, and does not even read from `dummy`.
+#[unstable(feature = "test", issue = "27812")]
+pub fn black_box<T>(dummy: T) -> T {
+    #[cfg(not(
+        any(
+            target_arch = "asmjs",
+            all(
+                target_arch = "wasm32",
+                target_os = "emscripten"
+            )
+        )
+    ))] {
+        // we need to "use" the argument in some way LLVM can't
+        // introspect.
+        unsafe { asm!("" : : "r"(&dummy)) }
+        dummy
+    }
+    #[cfg(
+        any(
+            target_arch = "asmjs",
+            all(
+                target_arch = "wasm32",
+                target_os = "emscripten"
+            )
+        )
+    )] {
+        // asm.js and emscripten do not support inline assembly
+        unsafe {
+            let ret = crate::ptr::read_volatile(&dummy);
+            crate::mem::forget(dummy);
+            ret
+        }
+    }
```

Comment:
To simplify the `#[cfg]` here, perhaps this could be:

```rust
pub fn black_box<T>(dummy: T) -> {
    #[cfg(supports_asm)]
    {
        unsafe { asm!(...); }
        return dummy;
    }
    unsafe {
        let ret = read_volatile(...);
        ...;
    }
}
```

Maybe with an `#[allow(unreachable_code)]` or something like that to pacify the linter?

Reply:
The last commit adds `cfg_if!` to the `internal_macros.rs` and uses it here. I find this looks infinitely better with `cfg_if!` than with any of the other options, and I suppose we could access it from `core::arch` as well.